### PR TITLE
Remove invalid UTF8 chars from OPDS

### LINKED ISSUE #2627: OPDS: Crash when downloading books with unknown characters in title
* KOReader version: last development version
* Device: Kobo Mini

#### Issue
Downloading books from ManyBooks work if you don't get #2626 bug. But here we have a crash if the title of the book to download has an unknown character. You will be able to select the book, and begin the download, but before it finishes, KOReader will crash and close itself, exiting to KSM or Nickel.

#### Steps to reproduce
Any title with the unknown character � will crash the device. You can see one going to
OPDS Catalog->ManyBooks->Genres->Classic and you will see there Be�wulf. Select it, begin the download and after some seconds, KOReader will crash and close.

[crash.log.txt](https://github.com/koreader/koreader/files/834188/crash.log.txt)

### PR DESCRIPTION
**Pull request by robert00s**
Should fix #2627 and should better work than #2628 
I found this function http://notebook.kulchenko.com/programming/fixing-malformed-utf8-in-lua that should remove invalid UTF-8 chars.

### CODE REVIEW

**Comment by Frenzie** [comment]
It'd be cool if you could also add a unit test to help prevent breaking this function in the future. That's in https://github.com/koreader/koreader/blob/10132d1c5972befd4e3e3acd0f4d3c3a4952d12c/spec/unit/util_spec.lua

I'm talking about something like this, except you'll have to double check if I:

1. wrote this correctly at all
2. perhaps also come up with a better test (e.g., based on whatever issue you ran into that required this in the first place)

```lua
    it("should replace invalid UTF-8 characters with an underscore", function()
        assert.is_equal(util.fixUtf8("\127 \128 \194\127 ", "_"), "\127 _ _\127_")
    end)
```

File: frontend/ui/widget/opdsbrowser.lua [hunk 1]
```diff
@@ -20,6 +20,7 @@ local http = require('socket.http')
 local https = require('ssl.https')
 local ltn12 = require('ltn12')
 local mime = require('mime')
+local util = require("util")
```
Comment (Frenzie) [inline frontend/ui/widget/opdsbrowser.lua]: Oh, and if you happen to have the time there's a new part of the coding style about the order of requires: https://github.com/koreader/koreader-base/wiki/Coding-style#order-of-requires

But only if you feel like it. ;-)

File: frontend/util.lua [hunk 2]
```diff
@@ -299,4 +299,26 @@ function util.getMenuText(item)
...
+        elseif pos == str:find(       "\224[\160-\191][\128-\191]", pos)
+            or pos == str:find("[\225-\236][\128-\191][\128-\191]", pos)
+            or pos == str:find(       "\237[\128-\159][\128-\191]", pos)
+            or pos == str:find("[\238-\239][\128-\191][\128-\191]", pos) then pos = pos + 3
+        elseif pos == str:find(       "\240[\144-\191][\128-\191][\128-\191]", pos)
+            or pos == str:find("[\241-\243][\128-\191][\128-\191][\128-\191]", pos)
+            or pos == str:find(       "\244[\128-\143][\128-\191][\128-\191]", pos) then pos = pos + 4
+        else
+            str = str:sub(1, pos - 1) .. replacement .. str:sub(pos + 1)
+            pos = pos + 1
```
Comment (Frenzie) [inline frontend/util.lua]: replacement can be more than one character, so you need to do `pos = pos + #replacement`

Comment (Frenzie) [inline frontend/util.lua]: Oh, and that also means `len = len + #replacement`

File: frontend/util.lua [hunk 3]
```diff
@@ -299,4 +299,26 @@ function util.getMenuText(item)
     return text
 end
 
+-- from http://notebook.kulchenko.com/programming/fixing-malformed-utf8-in-lua with modification
```
Comment (Frenzie) [inline frontend/util.lua]: Could you also add docs?

```lua
--- Replaces invalid UTF-8 characters with a replacement string.
---- @string str the string to be checked for invalid characters
---- @string replacement the string to replace invalid characters with
---- @treturn string valid UTF-8
```

File: frontend/util.lua (hunk 2 above)
Comment (Frenzie) [inline frontend/util.lua]: That also brings up a second unit test:
```
    it("should replace invalid UTF-8 characters with multiple characters", function()
        assert.is_equal(util.fixUtf8("\127 \128 \194\127 ", "__"), "\127 __ __\127 ")
    end)
```

**Comment by Frenzie** [comment]
Sorry for being a bit overbearing on this one, but it means I like it! ;-D

**Comment by robert00s** [comment]
@Frenzie No problemo :)
Thanks for help.

**Pull request merged**